timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/TimelockLeadershipMetrics.java

```diff
@@ -16,58 +16,120 @@
+    @Value.Derived
+    LeadershipObserver leadershipObserver() {
+        return leadershipObserverFactory().create(client());
+    }
+
+    public <T> T instrument(String name, Class<T> clazz, T instance) {
         return AtlasDbMetrics.instrumentWithTaggedMetrics(
                 metrics().metrics(),
                 clazz,
                 instance,
                 name,
                 _context -> ImmutableMap.of(
-                        AtlasDbMetricNames.TAG_CLIENT, client.value(),
-                        AtlasDbMetricNames.TAG_CURRENT_SUSPECTED_LEADER, String.valueOf(localPingableLeader().ping())));
+                        TAG_CLIENT, client().value(),
+                        TAG_CURRENT_SUSPECTED_LEADER, String.valueOf(localPingableLeader().ping())));
     }
 
-    @Value.Derived
-    public List<SafeArg<Object>> namespaceAsArgs() {
-        return ImmutableList.of(SafeArg.of(AtlasDbMetricNames.TAG_PAXOS_USE_CASE, metrics().paxosUseCase().toString()));
+    public static AutobatchingLeadershipObserverFactory createFactory(TimelockPaxosMetrics metrics) {
+        return AutobatchingLeadershipObserverFactory.create(getMetricDeregistrator(metrics));
+    }
+
+    private static Consumer<SetMultimap<LeadershipEvent, Client>> getMetricDeregistrator(TimelockPaxosMetrics metrics) {
+        PaxosUseCase paxosUseCase = metrics.paxosUseCase();
+        switch (paxosUseCase) {
+            case LEADER_FOR_ALL_CLIENTS:
+                return deregisterMetricsForSingleLeader(metrics);
+            case LEADER_FOR_EACH_CLIENT:
+                return deregisterMetricsForPartitionedLeader(metrics);
+            case TIMESTAMP:
+                throw new SafeIllegalArgumentException("Timestamp paxos not supported");
+            default:
+                throw new SafeIllegalArgumentException("Unexpected paxosUseCase",  SafeArg.of("useCase", paxosUseCase));
+        }
+    }
+
+    private static Consumer<SetMultimap<LeadershipEvent, Client>> deregisterMetricsForSingleLeader(
+            TimelockPaxosMetrics metrics) {
+        return eventsToDeregister -> eventsToDeregister.keySet().forEach(event -> metrics.asMetricsManager()
+                .deregisterTaggedMetrics(withTagIsCurrentSuspectedLeader(event.isCurrentSuspectedLeader())));
+    }
+
+    private static Consumer<SetMultimap<LeadershipEvent, Client>> deregisterMetricsForPartitionedLeader(
+            TimelockPaxosMetrics metrics) {
+        return eventsToDeregister -> {
+            Map<LeadershipEvent, Set<String>> eventsToClients = Multimaps.asMap(KeyedStream.stream(eventsToDeregister)
+                    .map(Client::value)
+                    .collectToSetMultimap());
+            eventsToClients.keySet().forEach(event -> metrics.asMetricsManager().deregisterTaggedMetrics(
+                    withTagIsCurrentSuspectedLeader(event.isCurrentSuspectedLeader())
+                            .and(withClientTag(eventsToClients.get(event)))));
```

Comment:
maybe use `entrySet`? Not superhot codepath so I'm not too strongly concerned

Reply:
with entry set, run time becomes `#metrics * #clients`, since deregister will enumerate each metric and run the predicate, can't get rid of the `#metrics`

Reply:
spoke offline, entry set of `Map<Bla, Collection<Bla2>>`